### DISCUSSION TITLE
Alternative to type_ and ModelMetaclass

### DISCUSSION THREAD

**Discussion by deivydaslipskis** [pydantic/pydantic#7218] [Pydantic V2]
In Pydantic v1 I would use `type_` on `ModelField` and check if it is an instance of `ModelMetaclass`, how can I do the same thing in Pydantic v2?

Example:
```
print(model.__dict__["__fields__"])
{'id': ModelField(name='id', type=str, required=True), 'name': ModelField(name='name', type=str, required=True), 'is_unicorn': ModelField(name='is_unicorn', type=bool, required=True), 'is_claimed': ModelField(name='is_claimed', type=bool, required=True), 'ceo': ModelField(name='ceo', type=Person, required=True), 'employees': ModelField(name='employees', type=List[Person], required=True), 'hq': ModelField(name='hq', type=Optional[Building], required=False, default=None), 'rating': ModelField(name='rating', type=CompanyRating, required=True)}
```

I would check it like this:

```
for key, value in model.__dict__["__fields__"].items():
    if isinstance(value.type_, ModelMetaclass):
        # do something
        # This should catch ceo, employees, hq, rating fields

```
In Pydantic v2 accesing fields has changed and the values inside the dict have also changed:

```
print(model.model_fields)
`{'id': FieldInfo(annotation=str, required=True), 'name': FieldInfo(annotation=str, required=True), 'is_unicorn': FieldInfo(annotation=bool, required=True), 'is_claimed': FieldInfo(annotation=bool, required=True), 'ceo': FieldInfo(annotation=Person, required=True), 'employees': FieldInfo(annotation=List[Person], required=True), 'hq': FieldInfo(annotation=Union[Building, NoneType], required=True), 'rating': FieldInfo(annotation=CompanyRating, required=True)}`
```
The `ModelMetaclass` has been deprecated also.

How do I check the type if the `type_` method and `ModelMetaclass` have been deprecated?

**Comment by harunyasar** [comment]
In fact, V1 and V2 both has the `annotation` attribute. So, you could use

```python
m.__fields__[field_name].annotation.__origin__
m.__fields__[field_name].annotation.__args__  # gives you the generic type
```

**Reply by deivydaslipskis** [reply]
Thank you for your answer.
This seems to work only with the `List` type, all others fail in my case. When using these methods on other types I get these errors:
When using `__args__` receiving: `*** AttributeError: __args__ or` and `*** AttributeError: type object 'str' has no attribute '__args__'. Did you mean: '__add__'?`
And when using `__origin__`: `*** AttributeError: __origin__` and `*** AttributeError: type object 'str' has no attribute '__origin__'`

Here is an extended example:

```
from pydantic import BaseModel
from typing import Optional, get_origin, List


class Name(BaseModel):
    """
    BaseModel subclass
    """


class Building(Name):
    address: str
    officers: list[str]


class Person(Name):
    firstname: str
    lastname: Optional[str]


class Company(Name):
    id: str
    hq: Optional[Building]
    name: str
    is_claimed: bool
    ceo: Person
    employees: List[Person]
    slug: Optional[str]


model = Company(
    id="123",
    hq=Building(address="test1", officers=["test2"]),
    name="my_company",
    ceo=Person(firstname="test3", lastname="test4"),
    is_claimed=True,
    employees=[
        Person(firstname="test3", lastname="test4"),
        Person(firstname="test5", lastname="test6"),
    ],
    slug="comp",
)

#model_fields print:
#{'id': FieldInfo(annotation=str, required=True), 'hq': FieldInfo(annotation=Union[Building, NoneType], required=True), 'name': FieldInfo(annotation=str, required=True), 'is_claimed': FieldInfo(annotation=bool, required=True), 'ceo': FieldInfo(annotation=Person, required=True), 'employees': FieldInfo(annotation=List[Person], required=True), 'slug': FieldInfo(annotation=Union[str, NoneType], required=True)}


for key, value in model.model_fields.items():
    if isinstance(value.annotation, BaseModel):
        print(f"Key - {key} is instance of BaseModel")
        # should print on hq, ceo keys

    if isinstance(value.annotation, Name):
        print(f"Key - {key} is instance of Name")
        # should print on hq, ceo keys

    if get_origin(value.annotation) == list:
        print(f"Key - {key} is list")
        # this seems to work and does not error on other types

    # if issubclass(value.annotation, BaseModel):
    #     TypeError: issubclass() arg 1 must be a class

    # if issubclass(value.annotation, Name):
    #     TypeError: issubclass() arg 1 must be a class


#How we were using this in Pydantic v1:

from pydantic.main import ModelMetaclass

for key, value in model.__dict__["__fields__"].items():
            if isinstance(value.type_, ModelMetaclass):
                # do something with hq, ceo
            elif get_origin(value.outer_type_) == list:
                # do something with employees
            else:
                # do something with leftover types like int, str, bool

```
I can handle `List` types with `get_origin()` but I am still not sure how to check if the type is an instance of the previous `ModelMetaClass` (or its alternative or at least `BaseModel`).

In short, I need a way to discern if the type is an instance of the previous `ModelMetaclass` (or its alternative or `BaseModel`).

**Reply by harunyasar** [reply]
I know what you mean but with `annotation` attribute you could achieve the same thing. For example, `str` field won't have a `__origin__` attribute you could simply check the type of the `annotation` then. Since I can't see a direct replacement for these attributes you are supposed to write your own logic by looking at what `get_origin` and `get_args` return. This is what I understand.

**Reply by deivydaslipskis** [reply]
got it, thanks
